sdc/rewrites/read_csv_consts.py

```diff
@@ -0,0 +1,161 @@
+    var_def = get_definition(func_ir, var)
+    require(isinstance(var_def, ir.Expr))
+    build_ops = ['build_tuple', 'build_list', 'build_set', 'build_map']
+    require(var_def.op in build_ops)
+    return var_def.items, var_def.op
+
+
+def _new_definition(func_ir, var, value, loc):
+    func_ir._definitions[var.name] = [value]
+    return ir.Assign(value=value, target=var, loc=loc)
+
+
+class ConstantInference(consts.ConstantInference):
+
+    def _infer_expr(self, expr):
+        if expr.op == 'build_map':
+            def inf_const(value):
+                return self.infer_constant(value.name, loc=expr.loc)
+            return {inf_const(k): inf_const(v) for k, v in expr.items}
+        return super()._infer_expr(expr)
+
+
+@register_rewrite('before-inference')
+class RewriteReadCsv(Rewrite):
+    """
+    Searches for calls of pandas.read_csv() and replace it with calls of read_csv.
+    """
+
+    _pandas_read_csv_calls = [
+        ('read_csv', 'pandas'),             # for calls like pandas.read_csv()
+        ('read_csv', 'pandas.io.parsers'),  # for calls like read_csv = pandas.read_csv, read_csv()
+    ]
+
+    _read_csv_const_args = ('names', 'dtype', 'usecols')
+
+    def match(self, func_ir, block, typemap, calltypes):
+        # TODO: 1. save instructions of build_map, build_list for read_csv params
+        # 2. check that vars are used only in read_csv
+        # 3. replace vars with build_tuple inplace
+
+        self.func_ir = func_ir
+        self.block = block
+        self.consts = consts = {}
+
+        # Find all assignments with a right-hand read_csv() call
+        for inst in block.find_insts(ir.Assign):
+            if not isinstance(inst.value, ir.Expr):
+                continue
+            expr = inst.value
+            if expr.op != 'call':
+                continue
+            call = guard(find_callname, func_ir, expr)
+            if call not in self._pandas_read_csv_calls:
+                continue
+            # protect from repeat rewriting
+            if hasattr(inst, 'consts'):
+                continue
+            # collect constant parameters
+            for key, var in expr.kws:
+                if key not in self._read_csv_const_args:
+                    continue
+                try:
+                    const = func_ir.infer_constant(var)
+                except errors.ConstantInferenceError:
+                    try:
+                        const = ConstantInference(func_ir).infer_constant(var.name)
+                    except errors.ConstantInferenceError:
+                        continue
+                consts.setdefault(inst, {})[key] = const
+
+        return len(consts) > 0
+
+    def apply(self):
+        new_block = self.block.copy()
+        new_block.clear()
+        vars_to_remove = []
+
+        for inst in self.block.body:
+            if inst in self.consts:
+                # protect from repeat rewriting
+                inst.consts = consts = self.consts[inst]
```

Comment:
alternatively you could in `match` check that specific parameter is actually a dict, and if it is not - do not trigger `match`

Reply:
Fixed. Use checking.